### DISCUSSION TITLE
Architecture options

### DISCUSSION THREAD

**Discussion by crazy4pi314** [Stage-Whisper/Stage-Whisper#33] [General]
Past the first MVP, we would like to be able to scale the architecture of the app so that it can support:
- mobile interfaces
- Parallel model execution
- Parallel file processing
- ... and probably more (see [ideas](https://github.com/Stage-Whisper/Stage-Whisper/discussions/categories/ideas) discussion :) )

So having a messaging service or even a "UI call the Python CLI" app model will probably limit our options. I have been doing some research and consulting other devs and it seems like a microservices architecture could be nice here. The idea would be as follows:

<img width="1148" alt="image" src="https://user-images.githubusercontent.com/6486256/193331203-af19e92b-5f97-4535-ae41-79165a865e02.png">

*Selected technologies for each component are options, not requirements.

We would be able to use the nice Electron app that @oenu has been working on, paired with some other threads running the Python models, and the coordination, state management, and security handled by Dapr.

This makes it easy to separate responsibilities, coordinate front and back end without thread locking, and allow for service discovery (say you have additional ML models or maybe a phone you want to add to the communication network.

This makes an interesting story for a mobile experience, where no heavy lifting would have to happen on a phone, and basically a lightweight webapp on the phone side could securely transfer (stream?) audio to the running desktop app for analysis. An example of adding a mobile interface: 

<img width="1173" alt="image" src="https://user-images.githubusercontent.com/6486256/193332204-b8f7fc79-c7b9-4385-ac19-7ec290a20597.png">

Again this is all pretty rough, but would make it pretty easy to scale and decouple development for different parts. One other additional benefit of using this sort of architecture is that if there was a cloud service that you wanted to include (say AWS transcribe or Azure Cognitive services) you would be able to have that "discovered" by Dapr and displayed in the electron app.

An example of this could be you from the desktop client want to transcribe a public video/audio somewhere on the internet, you could have a cloud microservice like AWS lambda or Azure functions that Dapr would send the URL from the user interface too, and the cloud microservice would download the media, run whatever models in the cloud and send back to Dapr (and thus the UI) the result.

<img width="509" alt="image" src="https://user-images.githubusercontent.com/6486256/193334119-f956c6e6-2095-4174-9a3b-f009b3f5be2b.png">

Would love some feedback on the idea, and any other propositions as I think this is a pretty wide open field :)

**Comment by petersterne** [comment]
Intriguing! I really like the flexibility. Just a few questions:

First, could this work entirely with local services on-device (in addition to being potentially extended to cloud services later on)? I want to make sure we maintain the security advantage of on-device transcription. 

Second, would a dapr-based model make it easier or harder to package and distribute the desktop executables?

Third, do you or any of our current developers have experience with dapr-based microservices, or do you think the devs you've been speaking to might be willing to contribute to the project? I'm just a bit wary of planning an ambitious architecture plan that no one on the team has experience working with.

**Comment by crazy4pi314** [comment]
> First, could this work entirely with local services on-device (in addition to being potentially extended to cloud services later on)? I want to make sure we maintain the security advantage of on-device transcription.

Yup, I would have this run entirely locally by default. It would have to be an opt-in/separate setup to do this with any cloud stuff.

> Second, would a dapr-based model make it easier or harder to package and distribute the desktop executables?

Learning now :) I don't think it is harder per say, but yeah

> Third, do you or any of our current developers have experience with dapr-based microservices, or do you think the devs you've been speaking to might be willing to contribute to the project? I'm just a bit wary of planning an ambitious architecture plan that no one on the team has experience working with.

I don't have dapr experience, but some folks who do who are interested. I totally agree about not over engineering here, I am really just trying to explore options. Def don't want to blow up scope, and not trying to say I am super attached to this idea. I just am trying things out :)

**Reply by crazy4pi314** [reply]
Also wanted to add a nice blog covering how Dapr works, and an example of local setup: https://medium.com/c-sharp-progarmming/building-distributed-apps-using-dapr-locally-and-on-azure-65316a2f259
